docs/codeql/codeql-language-guides/customizing-library-models-for-java.rst

```diff
@@ -0,0 +1,439 @@
+The first five values identify the callable (in this case a method) to be modeled as a sink.
+
+- The first value **java.sql** is the package name.
+- The second value **Statement** is the name of the class (type) that contains the method.
+- The third value **True** is a flag that indicates whether or not the sink also applies to all overrides of the method.
+- The fourth value **execute** is the method name.
+- The fifth value **(String)** is the method input type signature.
+
+The sixth value is only relevant internally and can be omitted in most use cases.
+The remaining values are used to define the **access path**, the **kind**, and the **provenance** (origin) of the sink.
+
+- The seventh value **Argument[0]** is the **access path** to the first argument passed to the method, which means that this is the location of the sink.
+- The eighth value **sql** is the kind of the sink. The sink kind is used to define the queries where the sink is in scope. In this case - the SQL injection queries.
+- The ninth value **manual** is the provenance of the sink, which is used to identify the origin of the sink.
+ 
+Example: Taint source from the **java.net** package.
+----------------------------------------------------
+In this example we show how to model the return value from the **getInputStream** method as a **remote** source.
+This is the **getInputStream** method in the **Socket** class, which is located in the **java.net** package.
+Please note that this source is already added to the CodeQL Java analysis.
+
+.. code-block:: java
+
+   public static void tainted(Socket socket) throws IOException {
+       InputStream stream = socket.getInputStream(); // The return value of this method is a remote source of taint.
+       ...
+   }
+
+This means that we want to add a tuple to the **sourceModel**\(package, type, subtypes, name, signature, ext, output, kind, provenance) extensible predicate, which can be achieved by adding the following to a data extension file:
+
+.. code-block:: yaml
+
+   extensions:
+     - addsTo:
+         pack: codeql/java-all
+         extensible: sourceModel
+       data:
+         - ["java.net", "Socket", False, "getInputStream", "()", "", "ReturnValue", "remote", "manual"]
+
+Reasoning:
+
+Since we are adding a new source, we need to add a tuple to the **sourceModel** extensible predicate.
+The first five values identify the callable (in this case a method) to be modeled as a source.
+
+- The first value **java.net** is the package name.
+- The second value **Socket** is the name of the class (type) that contains the source.
+- The third value **False** is a flag that indicates whether or not the source also applies to all overrides of the method.
+- The fourth value **getInputStream** is the method name.
+- The fifth value **()** is the method input type signature.
+
+The sixth value is only relevant internally and can be omitted in most use cases.
```

Comment:
Same comment as above but ... It might be worth clarifying what you mean by "only relevant internally" but I may be betraying my layman status here so feel free to ignore this